src/persist-client/src/internal/state_versions.rs

```diff
@@ -620,19 +625,87 @@ impl StateVersions {
             "add_and_remove_rollups should apply to the empty state"
         );
 
-        let rollup = self.encode_rollup_blob(shard_metrics, &initial_state, rollup.key);
+        let rollup = self.encode_rollup_blob(shard_metrics, &initial_state, vec![], rollup.key);
         let () = self.write_rollup_blob(&rollup).await;
         assert_eq!(initial_state.seqno, rollup.seqno);
 
         let diff = StateDiff::from_diff(&empty_state.state, &initial_state.state);
         (initial_state, diff)
     }
 
-    /// Encodes the given state as a rollup to be written to the specified key.
+    pub async fn write_rollup_for_state<K, V, T, D>(
+        &self,
+        shard_metrics: &ShardMetrics,
+        state: &TypedState<K, V, T, D>,
+        rollup_id: &RollupId,
+    ) -> Option<EncodedRollup>
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let (latest_rollup_seqno, _rollup) = state.latest_rollup();
+        let seqno = state.seqno();
+
+        let diffs: Vec<_> = self
+            .fetch_all_live_diffs_gt_seqno::<K, V, T, D>(&state.shard_id, *latest_rollup_seqno)
```

Comment:
previously we've discussed caching these diffs locally, rather than fetching them from Consensus. I found this variation to be so enormously easier to write and reason about that I'm pretty strongly in the camp that we run with this first, and then revisit the decision later if it is for some reason problematic. The main concern that comes to mind is CRDB traffic, but I think these scans will be near negligible load / we can tune the rollup frequency.

Reply:
yeah, it's fair to frame this as potential followup work. could you put a TODO please?